### DISCUSSION TITLE
[atom] types for `grammar-registry.js` based off of incorrect docs

### DISCUSSION THREAD

**Discussion by icecream17** [DefinitelyTyped/DefinitelyTyped#56297] [Issues with a @types package]
The docs for `GrammarRegistry` point to https://github.com/atom/first-mate/blob/v7.4.1/src/grammar-registry.coffee instead of https://github.com/atom/atom/blob/v1.40.0/src/grammar-registry.js

(Issue in docs repo: https://github.com/atom/flight-manual.atom.io/issues/736)
(Types: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/atom/src/grammar-registry.d.ts)
(Code: https://github.com/atom/atom/blob/v1.40.0/src/grammar-registry.js)

Unfortunately, the types for Atom seem to be based on what the docs say instead of what the code says (understandable).

## Breakdown of methods / type changes

<details>
<summary>Unaffected methods</summary>

There's a lot of methods in common

I didn't check the parameters

```ts
onDidAddGrammar
onDidUpdateGrammar
getGrammars
loadGrammar
loadGrammarSync
readGrammar
readGrammarSync
maintainLanguageMode
assignLanguageMode
autoAssignLanguageMode
selectGrammar
getGrammarScore
```

</details>

### In the types but not in the code

```ts
onDidRemoveGrammar
```

### In the types, and in the code, but not documented (see note):

```ts
grammarForScopeName
addGrammar
removeGrammar
removeGrammarForScopeName
decodeTokens
```

### In the code but not in the types

```ts
assignGrammar
getAssignedLanguageId
addInjectionPoint

// Deprecated methods
// Not sure whether to include these methods
grammarOverrideForPath
setGrammarOverrideForPath
clearGrammarOverrideForPath (deprecated in code, but that's not mentioned in the docs)
```

### note

Often methods have documentation-comments in the code, before the function, like:
```js
  // Extended: Get all the grammars in this registry.

  // Experimental: Specify a type of syntax node that may embed other languages.
```

But many methods also don't have documentation. _I think_ the online documentation depends on the code documentation, so by default they're not included (since the types were made from docs).
```js
  shouldUseTreeSitterParser(languageId) {
    return this.config.get('core.useTreeSitterParsers', {
      scope: new ScopeDescriptor({ scopes: [languageId] })
    });
  }
```

[a comment by an automated account is not shown]

**Comment by lierdakil** [comment]
The concession we made while maintaining the types is that anything not in the docs is "private". Sometimes we include these "private" fields anyway, but there's no reasonable way to track those really from version to version, so we generally try to avoid those.

So anyway, yes, the docs are considered the primary source of truth. Regrettable that those have drifted. Would you like to make a pull request? I'd still prefer to maintain the "undocumented=private" rule where possible, so please avoid adding too many _new_ undocumented methods. If not, I'll try to get to it _sometime_, but I've  been pretty busy lately, not sure when I'll be able to get to it.

**Reply by icecream17** [reply]
I made a draft pr here: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/56369
